workspaces/diff-engine/cli/src/main.rs

```diff
@@ -106,57 +109,110 @@ fn main() {
+            let diff_comp =
+              tokio::task::spawn_blocking::<_, Option<(Vec<InteractionDiffResult>, Tags)>>(move || {
+                let interaction_json =
+                  interaction_json_result.expect("can read interaction json line from stdin");
+                let TaggedInput(interaction, tags): TaggedInput<HttpInteraction> =
+                  match serde_json::from_str(&interaction_json) {
+                    Ok(tagged_interaction) => tagged_interaction,
+                    Err(parse_error) => {
+                      eprintln!("could not parse interaction json: {}", parse_error);
+                      return None;
+                    }
+                  };
+
+                Some((diff_interaction(&projection, interaction), tags))
+              });
+            //dbg!("waiting for results");
+            let results = diff_comp
               .await
-            {
-              panic!("could not write diff result to results channel");
-              // TODO: Find way to actually write error info
+              .expect("diffing of interaction should be successful");
+            //dbg!("got results");
+
+            if let Some((results, tags)) = results {
+              for result in results {
+                //dbg!(&result);
+                if let Err(_) = results_sender
+                  .send(ResultContainer::from((result, &tags)))
+                  .await
+                {
+                  panic!("could not write diff result to results channel");
+                  // TODO: Find way to actually write error info
+                }
+              }
             }
+
+            drop(diff_task_permit);
+          });
+
+          // dbg!("Scheduling task", diff_tasks.len() + 1);
+          diff_tasks.push(diff_task);
+
+          Ok::<_, JoinError>(())
+        },
+
+        // propagating errors / panics from the diff tasks
+        Some(res) = async {
+          if diff_tasks.is_empty() {
+            None
+          } else {
+            // dbg!("Waiting for a worker task to finish");
+            let (finished_diff_task_result, task_index, _) = future::select_all(&mut diff_tasks).await;
+            // dbg!("Task finished!", task_index);
+            diff_tasks.remove(task_index);
+
+            Some(finished_diff_task_result)
           }
+        } => {
+          res
+        },
+
+        // propagating errors / panics from the results manager
+        Err(err) = &mut results_manager => {
+          Err(err)
         }
+      };
 
-        drop(diff_task_permit);
-      });
+      // panic when scheduling, diff tasks 
+      next_action.expect("essential worker task panicked");
     }
 
     drop(results_sender);
-    results_manager.await.unwrap(); // make sure the results manager is done flushing
+
+    let completing_diff_tasks = future::try_join_all(diff_tasks);
+    try_join!(completing_diff_tasks, results_manager).expect("essential worker task panicked");
```

Comment:
If we've arrived at this point, `stdin` has been exhausted. While only awaiting `results_manager` _will_ prevent us from shutting down until all diff tasks are done (it won't complete until all `result_senders` have been dropped, of which each of a `diff_task` has one), there is _also_ the possibility of one of the diff tasks failing. Therefor, it's best to wait until they are all either completed, or one of these things fails.